### ISSUE TITLE
Roslyn doens't verify inferred nullability of variables passed in delegate constructors

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#38173]
**Version Used**: 

```
branch: release/dev16.4-preview1

commit aef5f7afb9208964afe7d0a2b4536b735649b347 (HEAD -> release/dev16.4-preview1, origin/release/dev16.4-preview1)
Merge: 7c5e8fdf0b c4895b82cd
Author: Petr Hou<C5><A1>ka <t-pehouk@microsoft.com>
Date:   Tue Aug 20 23:33:36 2019 -0700

    Merge pull request #37884 from petrroll/move-refa-to-helpers6

    Move some ComputeRefactoringsAsync to common helpers 6/n
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
class C
{
  static void Main()
  {
      M1("");
  }
  static void M1(string? s)
  {
    if (s == null) return;

    hell:
    var d1 = GetD(s);
    var d2 = new D<string>(d1);
    d2().ToString();
    s = null;
    goto hell;
  }

  static D<T> GetD<T>(T t) => () => t;
}

delegate T D<T>();
```

**Expected Behavior**:
` CS8621: Nullability of reference types in return type of 'string? D<string?>.Invoke()' doesn't match the target delegate 'D<string>'.` for `new D<string>(d1)`

**Actual Behavior**:
No warnings are reported. The program crashes at runtime with a `NullReferenceException`

**Comment by RikkiGibson** [comment]
It seems like we're doing something for top-level nullability which makes this kind of scenario work, but perhaps not for nested nullability? @cston do you have any idea why?

```cs
class C
{
    static void M2(string? s)
    {
        if (s == null) return;

        hell:
        var s1 = Get(s);
        s1.ToString(); // warns as expected
        s = null;
        goto hell;
    }
    
    static T Get<T>(T t) => t;
}
```

I would also expect to get the same behavior when replacing `hell: ... goto hell;` with `while (true) { ... }`

**Comment by TessenR** [comment]
@RikkiGibson 
Actually nested nullability is tracked as well as you can check with the following code:
```cs
#nullable enable
public class C<T> {
    public void M(string? s) {
        if (s == null) return;
        
        hell:
        var c1 = GetC(s);
        C<string> c2 = c1; // CS8619: Nullability of reference types in value of type 'C<string?>' doesn't match target type 'C<string>
        s = null;
        goto hell;
    }
    
    extern C<T> GetC<T>(T t);
}
```

There are however some strange bugs with nested nullability which I can't explain e.g. a code sample from this issue: https://github.com/dotnet/roslyn/issues/33295

Of course any kind of syntax that allows control flow jumps to already visited location with updated variables' states will do e.g. any cycle.
